### ISSUE TITLE
Request for private routing integration

### ISSUE THREAD

**Issue by JSavage42** [cerner/terra-core#1934]
# Feature Request

## Description
An option to activate a private routing option in the routing configuration. 

By private routing I mean routes that are protected behind authentication. If you visit a website deep link that you have not been authenticated to see you will be redirected to a login page. Once you have logged in you will be redirected back to your requested page. If you are logged in already you go straight to the requested page.

Would be nice in the routingconfig.js
```
content: {
    '/login':{
      path: '/login',
      component: {
        default: {
          componentClass: Wrapper,
          props: {
            content: LoginPage,
          },
        },
      },
    },
    '/foo': {
      path: '/foo',
      component: {
        default: {
          componentClass: Wrapper,
          props: {
            isPrivate: true,
            content: Bar,
          },
        },
      },
    },
```
`isPrivate defaults to false.`

In the code it would save the url the user was trying to access in state. Pass that to the login component which processes authentication and pushed the requested url after authentication. 

## Additional Context / Screenshots

1. Similar to this website's tutorial: 
    - https://tylermcginnis.com/react-router-protected-routes-authentication/

2. Project I worked on while learning React:
https://github.com/JSavage42/expensify-app
    - Authenticated with Firebase
    - Uses redux and react-router
3. The authentication service implementation would be up to the team developing the application, but should be provided the appropriate pointers to get and push that information to the Login function. 

## @ Mentions
@tbiethman

**Comment by JakeLaCombe** [comment]
This maybe better in terra-framework, but we will need to see more uses of authentication before we can make a full decision. For example: How would we want to deal with Role Based authentication?

**Comment by bjankord** [comment]
I think dealing with anything related to authentication is beyond the scope of terra-ui. I'm not really interested in taking on in this concern in the open source codebase.

**Comment by bjankord** [comment]
Closing in favor of https://github.com/cerner/terra-framework/issues/306
There is a possible solution that would enable custom routing mentioned on the issue. We'll be discussing ways to make custom routing easier in the future and be posting on the issue in terra-framework.
